Add KEP on a new taint-manager (#3901)
kubernetes/enhancements: keps/sig-scheduling/3902-decoupled-taint-manager/README.md

- [ ]  (R) KEP approvers have approved the KEP status as `implementable`
- [ ]  (R) Design details are appropriately documented
- [ ]  (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
    - [ ]  e2e Tests for all Beta API Operations (endpoints)
    - [ ]  (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
    - [ ]  (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ]  (R) Graduation criteria is in place
    - [ ]  (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ]  (R) Production readiness review completed
- [ ]  (R) Production readiness review approved
- [ ]  "Implementation History" section is up-to-date for milestone
- [ ]  User-facing documentation has been created in [kubernetes/website](https://git.k8s.io/website), for publication to [kubernetes.io](https://kubernetes.io/)

> - [ ]  Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Author: @Huang-Wei , would you suggest which boxes I should check? Thanks.

> - [ ]  Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Reviewer: Just check the boxes that are already completed. Like "Enhancement issue in release milestone", "Impl. history", etc.

### Goals

* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns, maintainability. This separation also enables cluster-admins to build custom `TaintManager`s and use them in their clusters.

### Non-Goals

The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.

* Extend and enhance the current built-in APIs for customization of node taint-based eviction, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.

> * Actual extension and enhancement  of the current `TaintManager`.

Author: Would it be better to have them as part of the Goals instead if we add "Do Not" ?

- Do not introduce extra maintenance burden or regression 
- Do not introduce incompatible behavior to NodeLifecycleController or TaintManager

> * Actual extension and enhancement  of the current `TaintManager`.

Reviewer: Nope, the bullets are under `Non-Goals`, so use a positive tone is correct.

> * Actual extension and enhancement  of the current `TaintManager`.

Author: Are you suggesting we put the two bullets under `Non-Goals`? It sounds weird to me. If they are Non-Goals, would it meany we don't care if the changes will introduce extra maintenance burden and incompatible behavior? Also, these requirements should be self-evident (needless to say) for most changes if not all. 

```
Introduce extra maintenance burden or regression after the code reorganization
Introduce incompatible behavior of NodeLifecycleController or TaintManager
```

> * Actual extension and enhancement  of the current `TaintManager`.

Reviewer: > Are you suggesting we put the two bullets under Non-Goals?

Yes.

> Also, these requirements should be self-evident (needless to say) for most changes if not all.

It can be debating. But IMO, code restructuring  in this KEP is non-trivial work, and hence may cause regression/incompatibility issue, so listing them as Non-Goals doesn't sound redundant.

## Proposal

### User Stories (Optional)

#### Story

While this split is more of a kubernetes developer focused change, an effect of this change is that it allows cluster-admins to develop custom `TaintManager` and disable the default `TaintManager`. 

The default `TaintManager`, because of the implementation rigidity, may not be suitable for every cluster operator. We expect users to come up with custom `TaintManager`s and this KEP may exacerbate the problem of `bring your own taint manager`. However, discussing the use-cases for extending `TaintManager` or custom `TaintManager`s is beyond the scope of this KEP. 

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController` and created and run if `enable-taint-manager` flag is true. 

> <p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>

Reviewer 2: `![Current Controllers](current-nodelifecycle-controllers.png)`

A new `NodeLifecycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.

A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.   

The creation and startup of the default taint-manager is performed by `kube-controller-manager`. A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or  roll back to the old `NodeLifecycleController`.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.

##### Prerequisite testing updates

Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.

#####  Unit tests
- `pkg/controller/taintmanager`: 2023-05-23 - 0%  
  - Enable and disable the new `TaintManager`
  - The new `TaintManager` acts on node taints properly
- `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0% 

>   - Use the split `TaintManager` or the old combined controller

Reviewer 2: here you should list the existing tests, so I'd expect here `pkg/controller/nodelifecycle` and its subdirs.

#####  Integration tests

>  No integration tests are required. This feature is better tested with e2e tests.  

Reviewer 2: Why? Integration tests allow you to have a more strict environment that you have a full control on.

Author: Sounds good.  Added integration tests.

##### E2E tests

- Verify passing the existing E2E and conformance tests using the split `TaintManager`.
- Verify the ability to disable the default `TaintManager`.

### Graduation Criteria

#### Alpha

Since there are no new API changes, we can skip Alpha and go to Beta directly.

#### Beta

* Support `kube-controller-manager` flag `--controllers=-taint-manager` to disable the default taint-manager.
* Unit and e2e tests completed and passed.
* Performance and scalability tests to verify there is non-negligible performance degradation in node taint-based eviction.
* Update documents to reflect the changes.

#### GA

* Fix all reported bugs if any.

### Upgrade / Downgrade Strategy

> We’ll need to upgrade `kube-controller-manager` and` node-lifecycle-controller` and add a new controller `taint-manager`. Accordingly, a downgrade will revert to the previous version of `kube-controller-manager` and `node-lifecycle-controller` and remove the new added `taint-manager` . There should be no impact on the other components during upgrades and downgrades since user-exposed behavior should stay the same except a new flag to enable/disable the default `taint-manager`. This change should be clearly communicated to users.

Reviewer 2: This doesn't make sense. kube-controller-manager is one process, which embeds several controllers (node-lifecycle-controller and taint-manager, when added). The upgrade and downgrade description here should point out that this only affects kcm, so there are no expected problems during upgrade and downgrade. 
Unless you start disabling the built-in taint-manager and using a custom built one, in which case you're approach will be different, but that case is again non-goal, so I'm not sure it's worth describing it here.

Reviewer: I think we can reword in the perspective of user's angle:

For a user that doesn't use this feature, no matter which kind of upgrade/downgrade path is performed, the eventual result should be that both "applying NoExecute taints on nodes upon conditions" and "evict pods upon NoExecute taints" function well. (they don't care about the underlying code restructured or not)

For a user that uses this feature, downgrade would mean (in-tree) taint-manager will be enabled, no matter it's previously enabled or not; upgrade would mean (in-tree) taint-manager will be enabled, but can be disabled. We don't need to mention out-of-tree taint-manager, as 1) that's no different from an external controller, and 2) nothing can stop a customer runs duplicated or faulty taint-manager.

Author: How about the following

A feature gate `SeparateTaintManager`  enables and disables the new feature. When the feature is turned on, a user can use `kube-controller-manager`'s flag to disable the default `TaintManager`.

Reviewer 3: I think the point is that what you are saying is part of the Design, or the question "how can the user disable this feature".
But it's not part of the upgrade/downgrade strategy.
The answer is "N/A" for this KEP

### Version Skew Strategy

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

*This section must be completed when targeting alpha to a release.*

* **How can this feature be enabled / disabled in a live cluster?**
    * Enabled:  turn on the new feature gate `SeparateTaintManager`
    * Disabled: turn off the new feature gate `SeparateTaintManager`
* **Does enabling the feature change any default behavior?**
    * No, the default node-life-cycle controller and taint-manager behavior will stay the same.
* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
    * Yes, we will need to rack back `kube-controller-manager` and `node-lifecycle-controller`. 
* **What happens if we re-enable the feature if it was previously rolled back?**

>     * Upgrade `kube-controller-manager` and `node-lifecycle-controller` and add the new `taint-manager`.

Reviewer 2: This has a similar problem like my other comment.

Author: Changed it to  "Use the feature gate `SeparateTaintManager` to enable or disable the feature."

* **Are there any tests for feature enablement/disablement?**

>     * Planned for Alpha release

Reviewer 2: Alpha or beta?

Author: Alpha

### Rollout, Upgrade and Rollback Planning

*This section must be completed when targeting beta graduation to a release.*

* **How can a rollout or rollback fail? Can it impact already running workloads?**
  A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or the old combined `NodeLifecycleController`.

* **What specific metrics should inform a rollback?**
  N/A

* **Were upgrade and rollback tested? Was the upgrade→downgrade→upgrade path tested?**
  They will be covered in unit and e2e tests.

* **Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?**
  No

### Monitoring Requirements

*This section must be completed when targeting beta graduation to a release.*

* **How can an operator determine if the feature is in use by workloads?**
    * Disable the default taint-manager
    * Create custom taint-based pod eviction
* **How can someone using this feature know that it is working for their instance?**

>     * Node taints and taint-based pod eviction work as usual.

Reviewer 2: This won't allow you to differentiate between old and new behavior. If all goes as planned admins and users should not see the difference, right?

* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**

>     * Taint-based pod eviction latency

Reviewer 2: Metric?

Author: Taint-based pod eviction latency (e.g., p99) should be less than x seconds.

The existing  kube-controller-manager flag `--controllers` can be used to optionally disable the `NoExecuteTaintManager`.

![New Controllers](new-controllers.png)

### Implementation

A new `NodeLifecycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.

A new `NoExecuteTaintManager` is created as a top-level controller managed by` kube-controller-manager`. Its implementation is based on the current taint-manager from `controller/nodelifecycle/taint-manager.go`.   

The creation and startup of the default taint-manager is performed by `kube-controller-manager`. A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or  roll back to the old `NodeLifecycleController`.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.

##### Prerequisite testing updates

Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.

#####  Unit tests

> - `pkg/controller/taintmanager`: 2023-05-23 - 0%  

Reviewer: There is no pkg `pkg/controller/taintmanager`. You meant `pkg/controller/nodelifecycle/scheduler`?

Author: I'm referring it to the proposed new `TaintManager`.

  - Enable and disable the new `TaintManager`
  - The new `TaintManager` acts on node taints properly
- `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0% 
  - Test new configuration  

> - `pkg/controller/nodelifecycle`, `pkg/controller/nodelifecycle/scheduler`: 2023-05-23 - 100% 

Reviewer: 100% doesn't look right:

<img width="910" alt="image" src="https://github.com/kubernetes/enhancements/assets/1425903/0d32aadb-edba-48ef-a6a8-93d32b453404">

I'd suggest putting separate pkg/controller/nodelifecycle and pkg/controller/nodelifecycle/scheduler

Author: Made the suggested change.

* Unit and e2e tests completed and passed.
* Performance and scalability tests to verify there is non-negligible performance degradation in node taint-based eviction.
* Update documents to reflect the changes.

#### GA

* Fix all reported bugs if any.

### Upgrade / Downgrade Strategy

A feature gate `SeparateTaintManager`  enables and disables the new feature. When the feature is turned on, a user can use `kube-controller-manager`'s flag to disable the default `TaintManager`.

### Version Skew Strategy

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

*This section must be completed when targeting alpha to a release.*

* **How can this feature be enabled / disabled in a live cluster?**
    * Enabled:  turn on the new feature gate `SeparateTaintManager`
    * Disabled: turn off the new feature gate `SeparateTaintManager`
* **Does enabling the feature change any default behavior?**
    * No, the default node-life-cycle controller and taint-manager behavior will stay the same.
* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**

>     * Yes, we will need to rack back `kube-controller-manager` and `node-lifecycle-controller`. 

Reviewer: ```suggestion
    * Yes. Once rolled back, `taint-manager` is enabled along with `node-lifecycle-controller`. 
```

* **What happens if we re-enable the feature if it was previously rolled back?**
    * The feature will work as usual.
* **Are there any tests for feature enablement/disablement?**

>     * Planned for Beta release

Reviewer: ```suggestion
    * Appropriate tests have been added in the integration tests. See [Integration tests](#integration-tests) for more details.
```

This separation not only improves code organization but also makes it easier to improve `TaintManager` or build custom `TaintManager`.

## Motivation
`NodeLifecycleController` combines two independent functions: 
   * Adding a pre-defined set of `NoExecute` taints to nodes based on the node conditions 
   * Performing pod eviction on `NoExecute` taints

Splitting the `NodeLifecycleController` based on the above functionalities will help to disentagle code and make future extensions to either component manageable.

### Goals

* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns, maintainability. This separation also enables cluster-admins to build custom `TaintManager`s and use them in their clusters.

### Non-Goals

* Improve and change the APIs to support extension and customization of the existing `TaintManager`.
* Actual extension and improvement of the current `TaintManager`.

## Proposal

### User Stories (Optional)

#### Story

> While this split is more of a kubernetes developer focused change, an unintended positive effect of this change is that it allows cluster-admins to develop custom `TaintManager`s and disable the default `TaintManager`. 

Reviewer 3: ```suggestion
While this split is more of a kubernetes developer focused change, an effect of this change is that it allows cluster-admins to develop custom `TaintManager`s and disable the default `TaintManager`. 
```

> The default `TaintManager`, because of the implementation rigidity, may not be suitable for every cluster operator. We except users to come up with custom `TaintManager`s and this KEP may exacerbate the problem of `bring your own taint manager`. However, discussing the use-cases for extending `TaintManager` or custom `TaintManger`s is beyond the scope of this KEP. 

Author: Typo of `expect`.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 

<p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>

### Proposed Controllers

The proposed design refactors `NodeLifecycleController` into two independent controllers, which are managed by `kube-controller-manager`.

1. `NodeLifecycleController` monitors node health and adds `NotReady` and `Unreachable` taints to nodes 
2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction

> The existing  kube-controller-manager flag `--controllers` will be used to optionally disable the `NoExecuteTaintManager`.

Reviewer 3: ```suggestion
The existing  kube-controller-manager flag `--controllers` can be used to optionally disable either controller.
```

<p align="center"><image src="new-controllers.png" width=60%></p>

> ### Implementation

Reviewer 3: Will there be a feature gate guarding the split? In other words, can operators use the combined controller, if they find any issues with the split, without having to rollback?

```
func startNoExecuteTaintManager(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {`
    if !controllerContext.ComponentConfig.NoExecuteTaintManager.EnableNoExecuteTaintManager {
        return nil, false, nil
    }

taintManager, err = taintmanager.NewNoExecuteTaintManager(ctx, kubeClient, nc.podLister, nc.nodeLister, nc.getPodsAssignedToNode)
        nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
            AddFunc: controllerutil.CreateAddNodeHandler(func(node *v1.Node) error {
                nc.taintManager.NodeUpdated(nil, node)
                return nil
            }),
            UpdateFunc: controllerutil.CreateUpdateNodeHandler(func(oldNode, newNode *v1.Node) error {
                nc.taintManager.NodeUpdated(oldNode, newNode)
                return nil
            }),
            DeleteFunc: controllerutil.CreateDeleteNodeHandler(func(node *v1.Node) error {
                nc.taintManager.NodeUpdated(node, nil)
                return nil
            }),
        })
    }
        if err != nil {
        return nil, true, fmt.Errorf("failed to start the taint manager: %v", err)
    }
  }  
```

> ### Test Plan

Reviewer 3: Please follow the template https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#test-plan

Author: Updated the test plan following the new template.

* Update documents to reflect the changes.

#### GA

* Fix all reported bugs if any.

### Upgrade / Downgrade Strategy

A feature gate `SeparateTaintManager`  enables and disables the new feature. When the feature is turned on, a user can use `kube-controller-manager`'s flag to disable the default `TaintManager`.

### Version Skew Strategy

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

*This section must be completed when targeting alpha to a release.*

* **How can this feature be enabled / disabled in a live cluster?**
    * Enabled:  turn on the new feature gate `SeparateTaintManager`
    * Disabled: turn off the new feature gate `SeparateTaintManager`
* **Does enabling the feature change any default behavior?**
    * No, the default node-life-cycle controller and taint-manager behavior will stay the same.
* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
    * Yes, we will need to rack back `kube-controller-manager` and `node-lifecycle-controller`. 
* **What happens if we re-enable the feature if it was previously rolled back?**

>     * Use the feature gate `SeparateTaintManager` to enable or disable the feature.

Reviewer 2: This still doesn't answer the question, what can happen if you re-enable the feature after a previous rollback? For example: you upgraded the cluster and started using this feature and something went wrong, so you rollback to previous version to get the problem resolved. Afterwards you're again turning it on? What can go wrong? 

In your case I think it's safe to say that there should be no unexpected problems.

Author: Yes, there should be no problem. Updated the proposal.
```
The new feature will work as usual. 
```

* **Are there any tests for feature enablement/disablement?**
    * Planned for Beta release

### Rollout, Upgrade and Rollback Planning

*This section must be completed when targeting beta graduation to a release.*

* **How can a rollout or rollback fail? Can it impact already running workloads?**
  A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or the old combined `NodeLifecycleController`.

* **What specific metrics should inform a rollback?**
  N/A

* **Were upgrade and rollback tested? Was the upgrade→downgrade→upgrade path tested?**
  They will be covered in unit and e2e tests.

* **Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?**
  No

### Monitoring Requirements

*This section must be completed when targeting beta graduation to a release.*

* **How can an operator determine if the feature is in use by workloads?**
    * Disable the default taint-manager

>     * Create custom taint-based pod eviction

Reviewer 2: Given the change is deep in the core, there's no easy way to know that, is it?

Author: If the feature gate `SeparateTaintManage`r used?

* **How can someone using this feature know that it is working for their instance?**
    * Node taints and taint-based pod eviction work as usual. Admins and users should not see any different behavior if everything works as planned.
* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**

>     * Taint-based pod eviction latency (e.g., P99) should be less than x seconds.

Reviewer 2: What is "x seconds"?

Author: How about `N/A`? This changes should have minimal performance impact on the pod eviction behavior.

* **What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?**
    * Taint-based pod eviction latency
* **Are there any missing metrics that would be useful to have to improve observability of this feature?**

>     * Taint-based pod eviction latency 

Reviewer 2: Can you point out the exact metric name?

Author: Will change it to `No` as this change should have minimal performance impact on the pod eviction behavior.

The non-goals are the following.
* Introduce extra maintenance burden or regression after the code reorganization
* Introduce incompatible behavior of `NodeLifecycleController` or `TaintManager`
* Extend and enhance the current built-in APIs for customization of node taint-based eviction, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
* Actual extension and enhancement  of the current `TaintManager`.

## Proposal

### User Stories (Optional)

While this split is for improving the code maintainability, an effect of this change is that it allows cluster-admins to extend and enhance the default `TaintManager` and even replace the default `TaintManager` with a custom implementation. However, discussing such use-cases for extending `TaintManager` or writing  custom `TaintManager` is beyond the scope of this KEP. 

#### Story

### Notes/Constraints/Caveats (Optional)

> ### Risks and Mitigations

Reviewer 4: I guess the risk here is that it will be harder to implement scenarios that need both features to be in the same controller. Just dreaming out loud those might be:

1. Fast notification between these two controllers to minimize time between taint being applied and eviction started. 
2. Cancellation of Pods eviction.
3. Pod eviction logic un-tainting the Node based on some signal received during the evicton.

I think mitigation for all of them would be the fact that we likely don't want this behavior built-in. Worth mentioning I think

Author: Good point! How about adding the following

"
Compared with a single combined controller, the risks of using two separate controllers include the following.
* Slightly increase the communication overhead from applying node taints to performing pod eviction.
* Make cancellation of pod eviction and un-tainting nodes harder.

It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs.
"

Reviewer 4: > It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs

I do not agree with the statement that built-in is necessarily simple and separation is the way to meet the advanced needs. Combined controller can be configureable. I think justification is more along the lines of cleaner design is better as we do not believe those scenarios are very important

Author: > > It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs
> 
> I do not agree with the statement that built-in is necessarily simple and separation is the way to meet the advanced needs. Combined controller can be configureable. I think justification is more along the lines of cleaner design is better as we do not believe those scenarios are very important

How about the following? 

`
While the performance impact of the split controllers is expected to be minimal, and pod eviction cancellation and un-tainting nodes are not common use cases, splitting  `NodeLifecycleController` promotes a cleaner design and enhances the overall maintainability and extensibility of in the long term.
`

Author: Thanks, Sergey. May I assume the KEP has been approved by you (sig-node)?

#####  Unit tests
- `pkg/controller/taintmanager`: 2023-06-03 - 0% (to add)
  - Enable and disable the new `TaintManager`
  - The new `TaintManager` acts on node taints properly
- `pkg/controller/apis/config/v1alpha1`: 2023-06-03 - 0% 
  - Test new configuration  
- `pkg/controller/nodelifecycle`: 2023-06-03 - 74.8%
  - Test the combined controller
- `pkg/controller/nodelifecycle/scheduler`: 2023-06-03 - 90%
  - Test the combined controller

#####  Integration tests
- Verify the ability to disable the default `TaintManager`.
- Verify the new `TaintManager`to act on node taints properly.

##### E2E tests

- Verify passing the existing E2E and conformance tests using the split `TaintManager`.

### Graduation Criteria

#### Alpha

Since there are no new API changes, we can skip Alpha and go to Beta directly.

#### Beta

* Support `kube-controller-manager` flag `--controllers=-taint-manager` to disable the default taint-manager.
* Unit and e2e tests completed and passed.
* Performance and scalability tests to verify there is non-negligible performance degradation in node taint-based eviction.
* Update documents to reflect the changes.

#### GA

* Fix all reported bugs if any.

> ### Upgrade / Downgrade Strategy

Reviewer 4: if somebody had the controller replaced, after the split the second one will stay? SO they will need to disable one more?

Author: For downgrade, disabling the new feature using the feature gate will automatically disable any custom controller and use the current and default combined controller.

### Non-Goals

The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.

* Introduce extra maintenance burden or regression after the code reorganization
* Introduce incompatible behavior of `NodeLifecycleController` or `TaintManager`
* Extend and enhance the current built-in APIs for customization of node taint-based eviction, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
* Actual extension and enhancement  of the current `TaintManager`.

## Proposal

### User Stories (Optional)

> #### Story

Reviewer 3: If the focus of the KEP is maintainability, we don't really need a story. This is confusing.

Author: > If the focus of the KEP is maintainability, we don't really need a story. The paragraphs below are adding confusion.

How about moving it out of `Story` as the a topic level note? 

"While this split is for improving the code maintainability, an effect of this change is that it allows cluster-admins to extend and enhance the default `TaintManager` and even replace the default `TaintManager` with a custom implementation. However, discussing such use-cases for extending `TaintManager` or writing  custom `TaintManager` is beyond the scope of this KEP."

Reviewer 3: that sounds like a better location.

While this split is more of a kubernetes developer focused change, an effect of this change is that it allows cluster-admins to develop custom `TaintManager` and disable the default `TaintManager`. 

The default `TaintManager`, because of the implementation rigidity, may not be suitable for every cluster operator. We expect users to come up with custom `TaintManager`s and this KEP may exacerbate the problem of `bring your own taint manager`. However, discussing the use-cases for extending `TaintManager` or custom `TaintManager`s is beyond the scope of this KEP. 

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

> In Kubernetes version 1.27 and earlier, the `NoExecuteTaintManager` component is included within the `NodeLifecycleController`. It is created and executed only if the `enable-taint-manager` flag is set to true. 

Author: OK, removed  the sentence "t is created and executed only if the `enable-taint-manager` flag is set to true. "

* **How can a rollout or rollback fail? Can it impact already running workloads?**
  A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or the old combined `NodeLifecycleController`.

* **What specific metrics should inform a rollback?**
  N/A

* **Were upgrade and rollback tested? Was the upgrade→downgrade→upgrade path tested?**
  They will be covered in unit and e2e tests.

* **Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?**
  No

### Monitoring Requirements

*This section must be completed when targeting beta graduation to a release.*

* **How can an operator determine if the feature is in use by workloads?**
    * It can be determined by if the feature gate `SeparateTaintManager` is used.
* **How can someone using this feature know that it is working for their instance?**
    * Node taints and taint-based pod eviction work as usual. Admins and users should not see any different behavior if everything works as planned.
* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**

>     * N/A as the change will have minimal impact on the performance and there's no service associated with it. 

Reviewer 2: Theoretically - yes it should not have impact, but how a cluster administrator can learn that? Think of both controllers like black-boxes and what would you look at from the outside that will give you assurance that the black-box still works as before the upgrade? See my other comment about queues' metrics

* **What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?**

>     * N/A

Reviewer 2: I'd at least expect the metrics for both controllers' queues to maintain current levels, and that should be explicitly pointed out in this proposal.

Author: How about "The performance of node taint-based eviction should remain the same level as before."?

- `pkg/controller/nodelifecycle`: 2023-06-03 - 74.8%
  - Test the combined controller
- `pkg/controller/nodelifecycle/scheduler`: 2023-06-03 - 90%
  - Test the combined controller

#####  Integration tests
- Verify the ability to disable the default `TaintManager`.
- Verify the new `TaintManager`to act on node taints properly.

##### E2E tests

- Verify passing the existing E2E and conformance tests using the split `TaintManager`.

### Graduation Criteria

#### Alpha

Since there are no new API changes, we can skip Alpha and go to Beta directly.

#### Beta

* Support `kube-controller-manager` flag `--controllers=-taint-manager` to disable the default taint-manager.
* Unit and e2e tests completed and passed.
* Performance and scalability tests to verify there is non-negligible performance degradation in node taint-based eviction.
* Update documents to reflect the changes.

#### GA

* Fix all reported bugs if any.

### Upgrade / Downgrade Strategy

A feature gate `SeparateTaintManager`  enables and disables the new feature. When the feature is turned on, a user can use `kube-controller-manager`'s flag to disable the default `TaintManager`.

### Version Skew Strategy

> ## Production Readiness Review Questionnaire

Reviewer 5: Please update to the new KEP template (release team will ask for it anyway):
https://github.com/kubernetes/enhancements/blame/master/keps/NNNN-kep-template/README.md#L430

Author: done

### Feature Enablement and Rollback

*This section must be completed when targeting alpha to a release.*

* **How can this feature be enabled / disabled in a live cluster?**
    * Enabled:  turn on the new feature gate `SeparateTaintManager`
    * Disabled: turn off the new feature gate `SeparateTaintManager`
* **Does enabling the feature change any default behavior?**
    * No, the default node-life-cycle controller and taint-manager behavior will stay the same.
* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
    * Yes. Once rolled back, `taint-manager` is enabled along with `node-lifecycle-controller` 
* **What happens if we re-enable the feature if it was previously rolled back?**
    * The feature will work as usual.
* **Are there any tests for feature enablement/disablement?**
    * Planned for Beta release

>     * Appropriate tests have been added in the integration tests. See [Integration tests](#integration-tests) for more details.

Reviewer 5: These tests haven't been added, rather they are planned, right?

That being said, if I'm reading correctly, those tests are rather checking if things work fine when  FG is enabled. They are not trying to test the enablement/disablement - being "run k8s with FG enabled, restart controller-manager with FG disabled and check if things still work as expected (and vice-versa)"

Which BTW is probably fine here, given this is in-memory feature (so no state is really stored here).
So I would be ok with saying sth like:
"No enablement/disablement tests are needed since this is in-memory feature and just regular tests with the feature being enabled/disabled do the job."

Author: updated as suggested.

### Rollout, Upgrade and Rollback Planning

*This section must be completed when targeting beta graduation to a release.*

* **How can a rollout or rollback fail? Can it impact already running workloads?**

>   A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or the old combined `NodeLifecycleController`.

Reviewer 5: Please answer the question. Can it impact already running workloads?

Author: Does this make sense?

"This is an opt-in feature, and it does not change any default behavior. Unless there is a bug in the implementation, a  rollout can not fail. If a rollout does fail, running workloads will not be evicted propertly on tainted nodes. We don't excpet a rollback can fail."

* **What specific metrics should inform a rollback?**

>   N/A

Reviewer 5: This isn't N/A for sure.

Some huge number of pod evictions? [Please add appropriate metric]

Author: How about "A significantly changing number of pod evictions and/or a substantial increase in pod eviction latency"?

* **Were upgrade and rollback tested? Was the upgrade→downgrade→upgrade path tested?**

>   They will be covered in unit and e2e tests.

Reviewer 5: Will they? We don't really have downgrade e2e tests running anywhere.

Please run them manually and describe the test case
[https://github.com/kubernetes/enhancements/pull/3658 is a nice example]

Author: Added "manually verify the rollback and upgrade-downgrade-upgrade path will pass the e2e testing"

Author: > The overall proposal makes sense to me (including going directly to Beta), but I have a bunch of comments about the PRR itself - PTAL

Thanks for the review, Wojciech! I've updated the doc to (hopefully) address the comments :-). Can you please take another look when you get a chance? Thanks again!

* **Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?**
  No

### Monitoring Requirements

*This section must be completed when targeting beta graduation to a release.*

* **How can an operator determine if the feature is in use by workloads?**
    * It can be determined by if the feature gate `SeparateTaintManager` is used.
* **How can someone using this feature know that it is working for their instance?**

>     * Node taints and taint-based pod eviction work as usual. Admins and users should not see any different behavior if everything works as planned.

Reviewer 5: How can I determine it as an operator running gazillions clusters?

Please specify a metric (number of pod evictions is probably a best one).

* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**
    * The performance of node taint-based eviction should remain the same level as before.
* **What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?**

>     * The metrics for both `NodeLifecycleController` and `TaintManager`'s queues should stay the same levels as before.

Reviewer 5: What metrics? Please be specific.

Author: How about "the number of pod evictions and pod eviction latency"?

    * No

### Dependencies

* **Does this feature depend on any specific services running in the cluster?** 
    * N/A

### Scalability

* **Will enabling / using this feature result in any new API calls?**
    * No
* **Will enabling / using this feature result in introducing new API types?** 
    * No
* **Will enabling / using this feature result in any new calls to the cloud provider?**
    * No
* **Will enabling / using this feature result in increasing size or count of the existing API objects?**
    * No
* **Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?**
    * No
* **Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?**
    * No

### Troubleshooting

* **How does this feature react if the API server and/or etcd is unavailable?**
    * The same behavior as the current version: fail to add taint and/or evict pods on tainted nodes 
* **What are other known failure modes?**
    * No
* **What steps should be taken if SLOs are not being met to determine the problem?**

>     * If the default or custom taint-manager is working or not. 

Reviewer 5: I don't understand it - can you please clarify?

Author: Updated to 

If the pod eviction latency increases significantly, validate if the communication between `NodeLifecycleController` and `TaintManager` works. If the number of pod evictions is abnormal, run tests to verify the TaintManager works properly.

<p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>

### Proposed Controllers

The proposed design refactors `NodeLifecycleController` into two independent controllers, which are managed by `kube-controller-manager`.

1. `NodeLifecycleController` monitors node health and adds `NotReady` and `Unreachable` taints to nodes 
2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction

The existing  kube-controller-manager flag `--controllers` can be used to optionally disable the `NoExecuteTaintManager`.

<p align="center"><image src="new-controllers.png" width=60%></p>

### Implementation

A new `NodeLifecycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.

A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.   

The creation and startup of the default taint-manager is performed by `kube-controller-manager` . 

> ```

Reviewer 3: I would remove the code snippet. It's not relevant in the KEP.

Author: OK, deleted it.

                return nil
            }),
            UpdateFunc: controllerutil.CreateUpdateNodeHandler(func(oldNode, newNode *v1.Node) error {
                nc.taintManager.NodeUpdated(oldNode, newNode)
                return nil
            }),
            DeleteFunc: controllerutil.CreateDeleteNodeHandler(func(node *v1.Node) error {
                nc.taintManager.NodeUpdated(node, nil)
                return nil
            }),
        })
    }
        if err != nil {
        return nil, true, fmt.Errorf("failed to start the taint manager: %v", err)
    }
  }  
```

A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or  roll back to the old `NodeLifecycleController`.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.

##### Prerequisite testing updates

Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.

#####  Unit tests
- `pkg/controller/taintmanager`: 2023-05-23 - 0%

> - `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0%

Reviewer 3: what is the coverage for the existing code?

Author: -  Enable and disable the new TaintManager
- The new TaintManager acts on node taints properly

#####  Integration tests
 No integration tests are required. This feature is better tested with e2e tests.  

##### E2E tests

- Verify passing the existing E2E and conformance tests using the split `TaintManager`.
- Verify the ability to disable the default `TaintManager`.

### Graduation Criteria

> #### Alpha

Reviewer 3: Since there are no new API fields, we could skip alpha and go to beta directly.
However, PRR reviewer might have a different opinion.

Author: Updated it.

* Support `kube-controller-manager` flag `--controllers=-taint-manager` to disable the default taint-manager.

> * Unit and integration tests completed and passed.

Reviewer 3: E2E?

Author: Fixed it.

#### Beta

* Performance and scalability tests to verify there is non-negligible performance degradation in node taint-based eviction.
* Update documents to reflect the changes.

#### GA

* Fix all reported bugs if any.

### Upgrade / Downgrade Strategy

We’ll need to upgrade `kube-controller-manager` and` node-lifecycle-controller` and add a new controller `taint-manager`. Accordingly, a downgrade will revert to the previous version of `kube-controller-manager` and `node-lifecycle-controller` and remove the new added `taint-manager` . There should be no impact on the other components during upgrades and downgrades since user-exposed behavior should stay the same except a new flag to enable/disable the default `taint-manager`. This change should be clearly communicated to users.

### Version Skew Strategy

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

*This section must be completed when targeting alpha to a release.*

> * **How can this feature be enabled / disabled in a live cluster?**

Reviewer 3: Mention the feature gate

Reviewer 5: Please update to the new KEP template (as mentioned above) - the feature-gate is then required explicitly :)

* There are increasing needs to customize and extend the default node taint-based pod evictions in practice.
  * Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including making calls to external systems and take actions on  `NoExecute` taints with other conditions such as unhealthy network. 

>   * Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.

Reviewer 3: This cannot be a motivation, as `enable-taint-manager=false` would still evict pods, just through a different (even less flexible) mechanism

Author: Yes, `descheduler` or other controllers can still evict pods, but `enable-taint-manager=false` will disable `NoExecute`-based pod eviction performed by the default `TaintManager`. We will remove it if you are strongly against it. 
 /cc @atosatto

Reviewer 3: I'm not talking about descheduler. I'm talking about kube-controller-manager itself. When you set `enable-taint-manager=false`, kcm can evict pods just based on Node conditions.

Reviewer 6: > When you set enable-taint-manager=false, kcm can evict pods just based on Node conditions.

@alculquicondor this is correct. However the functionality was deprecated and has been removed in https://github.com/kubernetes/kubernetes/pull/115840 as part of the effort to improve code-health in NodeLifeCycleController which this KEP is part of (as per your comment in https://github.com/kubernetes/enhancements/pull/3901#discussion_r11937486320).

The main benefit from de-coupling taint-based eviction from NodeLifeCycleController is that it would enable end users who have the possibility to control their control-plane configuration, to disable entirely taint-based eviction in their clusters without having to mutate all of the Pods via tolerations. 

While I do agree that there is room for extending the current built-in APIs to add (e.g. web-hooks) this should be part of a  different proposal and not in-scope of the current KEP which is purely focusing on re-organising the code.

Reviewer 3: That is correct, it's removed.

But the KEP here is trying to say that because we removed the flag, there is less control over eviction. That is not true, thus can't be part of the motivation.

Reviewer 3: > the main benefit from de-coupling taint-based eviction from NodeLifeCycleController is that it would enable end-users who have the possibility to control their control-plane configuration, to disable entirely taint-based eviction in their clusters without having to mutate all of the Pods via tolerations.

I disagree. Most end-users won't be able to disable the taint-based-eviction. And if they want to have some behavior similar to NoExecute taints with extra functionality, they still have to re-implement what kcm offers. So this is not a good extension mechanism.

> While I do agree that there is room for extending the current built-in APIs to add (e.g. web-hooks) this should be part of a different proposal and not in-scope of the current KEP which is purely focusing on re-organising the code.

I fully agree. That's what I have listed as one of the options to move forward with this KEP: to focus on re-organising.

Author: Updated the alternative with a focus on the code refactoring too. 

## Alternatives

Keeping the code as-is is an option, but it will make future extension of `TaintManager` harder and less flexible

  * The requirement of custom taint-based eviction is not unique. It exists already in Kubernetes, where we call it `FullyDistributionMode`. When all nodes are unhealthy, `NodeLifecycleController` chooses not to apply `NoExecute` taints properly. In a real-world case, the integrator may choose to define the disruption rate/budget, etc.

### Goals

> * Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to support flexible extension and customization of node taint-based eviction.

Reviewer 3: ```suggestion
* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns and maintainability.
```

Reviewer 3: The current wording can give the false impression that anyone can implement an extension, which is not true. Only users who have the ability to replace or configure control plane components can.

Author: Good point, we'll clarify that a only the cluster administrators can do it.

Reviewer 7: It will monitor these information changes. If a node has not updated its status for a long time, then we can assume that an exception has occurred in this node, and some fault tolerance processing is needed to safely remove the pod on this node. These Pods are rebuilt on other nodes.

    * It will require either injecting tolerations to every single pod via mutating webhooks and/or changing the manifests of all the running workloads.
    * It will still need an additional controller to really evict after coordinating with the external reservation/lease system. This will add additional overhead to toleration injections.
    * Custom mutating webhook may interfere with the admission plugin that applies the default tolerations if they are trying to apply tolerations for similar taints.
    * Toleration update in Webhooks can also cause conflicts with older version of pods especially during upgrades and we need to ensure a toleration update is valid.

* **Enhance the existing toleration APIs**
    * Introduce a new Toleration effect `EvictIfAllowed` similar to `PreferNoSchedule` with additional fields on each Toleration to specify a reference to another CR/API object for how to reach the external system that governs if eviction is allowed similar to webhooks. Similar to the communication between API Server and admission webooks, `NoExecTaintManager` and an external service can establish whether eviction for a given pod is allowed or not.

>     * Add RBAC for tolerations API on who can add what tolerations. Otherwise, a user can add tolerations to their workloads(pods) and make every pod land on the node to cause a noisy neighbor and even DoS problem in the worst case.

Reviewer 6: I think this is a separated problem than the one we are aiming to solve with this proposal. While there is definitely room for extending the Kubernetes API to support a more graceful form of `NoExecute` respecting the PodDisruptionBudget API (we had a session on this at the Contributor Summit in Amsterdam last week), the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour vs having to patch Kubernetes Controller Manager.

Author: > I think this is a separated problem than the one we are aiming to solve with this proposal. While there is definitely room for extending the Kubernetes API to support a more graceful form of `NoExecute` respecting the PodDisruptionBudget API (we had a session on this at the Contributor Summit in Amsterdam last week), the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour vs having to patch Kubernetes Controller Manager.

Added the comment to the discussion of alternatives.

Author: The use and enhancement of tolerations by injecting tolerations for NoExecute is a not viable at scale because of the factors described above.

Although extending the Kubernetes API to support a more graceful form of NoExecute respecting the PodDisruptionBudget API is possible, this proposal aims to enable operators to extend Kubernetes by providing custom logic for handling taint-based evictions. This approach changes the default behavior rather than requiring patching the Kubernetes Controller Manager.

Author: Splitting the components is not only for cleanup, but also a prerequisite for enhancing the pod eviction manager, whether it's the native or a custom controller.

Regarding your point about enhancing the default controller, we believe it would be helpful, but we also recognize that it may not be practical or even possible to have a general controller that meets all needs in different use cases, as described in the motivation and use cases.

We will make further clarification in the proposal. 

Thank you for taking the time to review our KEP and share your thoughts.

Reviewer 4: this is what was discussed at SIG Node meeting as well. It will be great to capture improvements planned or desired and pain points and explain how decoupling will help to address those.

Author: We summarize the benefits and improvements by decoupling `TaintManager` from `NodeLifecycleController`.

- A more logical code structure with two separate controllers for tainting nodes and evicting pods respectively. 
- Greater flexibility to support customized taint-based eviction to meet the specific needs of different workloads than the use of the toleration-based mechanism, which is less flexible and harder to customize.
- Ability to enable/disable the default `TaintManager` by setting or unsetting the "--controllers" flag of `kube-controller-manager`.

Reviewer 8: I think what @yuanchen8911  is suggesting is decoupling will make it easy for the future changes which are listed above. For example, once the split happens, it would be easier to make a HTTP call to external system which has more context and knowledge of the workloads from the taint controller(from an implementation standpoint).

Author: > > the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour.
> 
> You can already do that by disabling the controller.
> 
> I don't see any problem with splitting the existing controller into two. This is a rather welcomed cleanup.
> 
> I disagree that the motivation for splitting it is so that people can write their own. We should be looking for opportunities to enhance the existing controller and API so that people don't feel the need to fork it.

I agree splitting will enable to enhance and improve the existing controllers, but it will also provide opportunities for cluster admins to write their own taint managers like scheduler plugins.

In this KEP, we propose to 

* Decouple `TaintManager` that performs taint-based pod eviction from `NodeLifecycleController` and make them two separate controllers: `NodeLifecycleManager` to add taints to unhealthy nodes and `TaintManager` to perform pod deletion on nodes tainted with NoExecute effect.

* Allow users to opt-out from the taint-based eviction while still retaining the overall functionality provided by `NodeLifecycleController.`

## Motivation

* Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including making calls to external systems and take actions on  `NoExecute` taints with other conditions such as unhealthy network. 

> * Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.

Author: Yes, but without the flag, it will be extremely difficult to change the default pod eviction behavior. By splitting the flag, we can reuse the controller-manager's flag `--controllers` by adding or removing taintmanager.

* The requirement of custom taint-based eviction is not unique. It exists already in Kubernetes, where we call it `FullyDistributionMode`. When all nodes are unhealthy, `NodeLifecycleController` chooses not to apply `NoExecute` taints properly. In a real-world case, the integrator may choose to define the disruption rate/budget, etc.

> * `NodeLifecycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycleController`'s responsibility is primarily on Node unhealthy-zone, `TaintManager` is concerned with actioning the taints added by `NodeLifecycleController` by performing Pods eviction. De-coupling `NoExecuteTaintManager` from `NodeLifecycleController` in `kube-controller-manager` would allow to implement a single eviction controller to take care of general Pod eviction and deletion for different cases based on workload needs, including `NoExecute` taint-based pod eviction other than node `Unhealthy` and `Unreachable`.   

Reviewer 3: This should be the top motivation and, from my point-of-view, the only motivation.

Author: OK, moved the point to the top and restructured the motivations as
1. Splitting the controllers will improve maintainability and enable more flexible extension and customization of taint-based eviction.
2. There are increasing needs to extend and customize the default taint-manager in practice.

### Goals

* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager.

> * Using the `--controllers`  flag of `kube-controller-manager` to enable and disable the default `taint-manager` to support customizable taint-manager controller. 

Reviewer 3: IIUC, this is about any controller

Reviewer 8: Only https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R89 is relevant here

### Non-Goals

> * Any change to the behavior of the current taint-manager

Reviewer 8: Yup. Remove this from non-goals. This KEP is just for the split which improves the code maintenance.

> * Any change to the behavior of the current taint-manager

Author: Agree "`This KEP is just for the split which improves the code maintenance and makes future extensions easier`." .  The non-goal here means extension and customization of the default taint-manager behavior is out of the scope of the proposal.

#### Story 2

A cluster orchestrator can design and implement a centralized pod eviction manager that handles all pod eviction and deletions on different `NoExecute`  taints (`Unreachable`, `NotReady` and others) and any custom taints.

#### Story 3

A custom taint manager can make calls to other controllers or external systems to make a better decision whether, when and how to delete running pods on the tainted nodes (e.g. to prevent quorum loss and/or support different mechanism than Pod Disruption Budget to calculate applications availability).

#### Story 4

A pod may be stuck in terminating state for various reason, e.g.,  `kubelet` is not up and `AppServer` cannot reach `kubelet`, etc. The hanging deletion may prevent new pods starting because the persistent volume claim is not released. A custom taint-manager can perform a forceful deletion as needed.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

> In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 

Author: What we meant here is there will be no way to disable the default pod eviction behavior with the flag removed.

> In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 

Reviewer 3: It always has been, before or after the flag was removed.

> In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 

Author: Before the flag is removed, we can at least disable the default TaintManager.

> In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 

Reviewer 3: But then pods are immediately deleted based on unhealthy Node conditions, unless you disable the entire NodeLifecycleController.

See this code in 1.26 https://github.com/kubernetes/kubernetes/blob/4dfe380379bc9b3c39f8b041a95d12b4e6ac0cf3/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1329-L1335

My point is that we didn't lose the ability to "disable pod deletion" because we removed `enable-taint-manager`. That was not the purpose of the flag.

<p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>

### Proposed Controllers

The proposed design refactors `NodeLifecycleController` into two independent controllers, which are managed by `kube-controller-manager`.

1. `NodeLifecycleController` monitors node health and adds `NotReady` and `Unreachable` taints to nodes 
2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction

The existing  kube-controller-manager flag `--controllers` will be used to optionally disable the `NoExecuteTaintManager`.

<p align="center"><image src="new-controllers.png" width=60%></p>

### Implementation

A new `NodeLifecycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.

A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 

`// NoExecuteTaintManagerConfiguration contains elements describing NoExecuteTaintManagerConfiguration`

> `type NoExecuteTaintManagerConfiguration struct {`

Reviewer 3: why do we need this, in addition to the `--controllers` flag?

Author: You are right. We don't need any new flag. Removed it.

    * No
* **Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?**
    * No
* **Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?**
    * No

### Troubleshooting

* **How does this feature react if the API server and/or etcd is unavailable?**
    * The same behavior as the current version: fail to add taint and/or evict pods on tainted nodes 
* **What are other known failure modes?**
    * No
* **What steps should be taken if SLOs are not being met to determine the problem?**
    * If the default or custom taint-manager is working or not. 

## Implementation History

* 2023-03-06: Initial KEP published. 

## Drawbacks

## Alternatives

* **Leverage the existing tolerations for `NoExecute` taints**: operators opt-out from taint-based eviction by injecting tolerations for NoExecute taints. It has a few major disadvantages over an independent `TaintManager`.
    * It will require to either inject tolerations to all pods via mutating webhooks and/or changing the manifests of all the running workloads. 

>     * It will require either injecting tolerations to every single pod via mutating webhooks and/or changing the manifests of all the running workloads.

Reviewer 3: same point as above?

    * It will still need an additional controller to really evict after coordinating with the external reservation/lease system. This will add additional overhead to toleration injections.
    * Custom mutating webhook may interfere with the admission plugin that applies the default tolerations if they are trying to apply tolerations for similar taints.
    * Toleration update in Webhooks can also cause conflicts with older version of pods especially during upgrades and we need to ensure a toleration update is valid.

* **Enhance the existing toleration APIs**

>     * Introduce a new Toleration effect `EvictIfAllowed` similar to `PreferNoSchedule` with additional fields on each Toleration to specify a reference to another CR/API object for how to reach the external system that governs if eviction is allowed similar to webhooks. Similar to the communication between API Server and admission webooks, `NoExecTaintManager` and an external service can establish whether eviction for a given pod is allowed or not.

Reviewer 3: There are two APIs that allow extensibility of kubernetes without disabling default behavior:
- finalizers https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/
- scheduling gates https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/

Could you consider similar approaches?

I don't think the proposal in this paragraph is detailed enough, hence why it might not feel sufficient for your use case.

Reviewer 8: > Could you consider similar approaches?

The only closest thing that I can think of is scheduler extender(similar to entity which is outside of Kube system).

>  hence why it might not feel sufficient for your use case.

Which use-case are you referring to here? Splitting taint-manager?

Reviewer 3: Splitting the controllers is not a use-case. It's an implementation detail.

I'm not sure what use case the authors are trying to solve. I suppose it's close to story 1.

> The only closest thing that I can think of is scheduler extender

That goes along the way of "I want to fully override the component", which is the same as this proposal.

Reviewer 3: Right, that could be a valid avenue towards providing flexibility. However, not all users have access to the control plane to configure the webhook. So a more generic API or something more similar to finalizers could be more usable.

Reviewer 8: Agreed. I think the changes that @yuanchen8911 is suggesting are at the API level and he is also against having a webhook. However the implementation details may vary. Having said that, I guess we're debating about how we want to implement a gap in Kube in the alternate section of this KEP. The gap and how it needs to be addressed warrants another KEP. So, if you've any objections with the wording or the content(including the alternatives section), please let us know and @yuanchen8911 can make changes to move forward.

Reviewer 3: I already expressed my point: I'm in favor of the refactoring. I disagree with the Motivation and Stories. Given the current motivations, I consider that the Alternatives section hand-waves that improvements to the API wouldn't work without going into much detail.

Reviewer 8: I agree with your assessment on the motivation and use-cases section. 

However in the alternatives section, https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R301 the KEP states that the goal is not to change the API at this point of time. 

I believe https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R299 is relevant only to using webhooks not extending API. We feel that extending API is a good alternative but it is a larger change and it needs addressing few gaps that we mentioned. We will change the wording if that is what concerns you. Let me know if you have any other concerns.

Author: > I also disagree with that stance.

What would be your recommendation, Aldo?  Any specific things you want us to add to the KEP?

Reviewer 8: I guess, the point of contention here is `the side-effect of splitting the controllers is that it allows cluster operators to build custom TaintManagers and disable the default TaintManager`. @yuanchen8911 went ahead and explained why it is beneficial to split from that standpoint. As I mentioned earlier, we cannot seriously explore alternatives in this KEP as we may digress from the original point. So, better to cut short and explain this KEP on separation of concerns and tell the side-effect of this split is people can disable default `TaintManager`.

Reviewer 3: I like that approach better. It doesn't give false expectations to other people following the KEP.

Author: Updated the Non-Goals to clarify the code reorganizations the main goal of the KEP as below.

### Non-Goals

The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to the `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.

* Extend and enhance the current built-in APIs for `TaintManager` customization, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
* Actual extension and enhancement  of the current `TaintManager`.

> # KEP-3902:  Decouple Taint-based Pod Eviction from Node LifeCycle Controller 

Author: The issue number is https://github.com/kubernetes/enhancements/issues/3902.

    - [ ]  e2e Tests for all Beta API Operations (endpoints)
    - [ ]  (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
    - [ ]  (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ]  (R) Graduation criteria is in place
    - [ ]  (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ]  (R) Production readiness review completed
- [ ]  (R) Production readiness review approved
- [ ]  "Implementation History" section is up-to-date for milestone
- [ ]  User-facing documentation has been created in [kubernetes/website](https://git.k8s.io/website), for publication to [kubernetes.io](https://kubernetes.io/)
- [ ]  Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

> In Kubernetes, `NodeLifeCycleController` applies predefined `NoExecute` taints (e.g., `Unreachable`, `NotReady`) when the nodes are determined to be unhealthy. After the nodes get tainted, the `taint-manager` does its due diligence to start deleting running pods on those nodes based on `NoExecute` taints, which can be added by anyone.

Author: `taint-manager` --> `TaintManager`

* The requirement of custom taint-based eviction is not unique. It exists already in Kubernetes, where we call it `FullyDistributionMode`. When all nodes are unhealthy, `NodeLifeCycleController` chooses not to apply `NoExecute` taints properly. In a real-world case, the integrator may choose to define the disruption rate/budget, etc.
* `NodeCycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycle` controller responsibility is primarily on Node unhealthy-zone, `TaintManger` is concerned with actioning the taints added by `NodeLifecycle` controller by performing Pods eviction. De-coupling `NoExecuteTaintManager` from `NodeLifecycle` controller in `kube-controller-manager` would allow to implement a single eviction controller to take care of general Pod eviction and deletion for different cases based on workload needs, including `NoExecute` taint-based pod eviction other than node `Unhealthy` and `Unreachable`.   

> 

Reviewer: It'd be good add a Notes section here to explain that the term `TaintManger` and `NoExecuteTaintManager` will be used interchangeably in this doc.

Author: Added a note section.

`taint-manager`,  `TaintManger` and `NoExecuteTaintManager`  are used interchangeably in this doc.

### Goals

* Move taint-based eviction implementation out of `NodeLifeCycleController` into a separate and independent taint-manager.
* Using the existing `kube-controller-manager ` `--controllers`  flag to enable and disable the default `taint-manager` to support customizable taint-manager controller. 

### Non-Goals

* Any change to the behavior of the current taint-manager

## Proposal

### User Stories (Optional)

#### Story 1

An operator of stateful workloads that use local persistent volume can disable the default taint-manager and use a custom controller instead. When a node is tainted with `NoExecute` , the custom controller won’t delete the running stateful pods on the tainted node until the data is safe,  e.g., after the application manager successfully checkpoints the data in or migrates the data to a safe place. 

#### Story 2

> A cluster orchestrator can design and implement a centralized pod eviction manager that handles all pod eviction and deletions on different `NoExecute`  taints (`Unreachable`, `NotReady` and others) and any custom taints, e.g., acting on the conditions defined in `descheduler`, such as node overloading, etc.

Author: done.

**Story 3**

A custom taint manager can make calls to other controllers or external systems to make a better decision whether, when and how to delete running pods on the tainted nodes (e.g. to prevent quorum loss and/or support different mechanism than Pod Disruption Budget to calculate applications availability).

#### Story 4

> A pod may be stuck in terminating state for various reason, e.g.,  `kubelet` is not up and `AppServer` cannot reach `kubelet`, etc. The hanging deletion may prevent new pods starting because the persistent volume claim is not released. A custom taint-manager can perform a forceful deletion as needed.

Author: Typo, `APIServer`. Fixed it.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

> ### Current implementation

Reviewer: The term `current` should be avoided to use: https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-that-will-soon-be-out-of-date.

Author: Deleted the word `Current` and changed the description to `In version 1.27 and prior of Kubernetes, ...`

In this current version, `NoExecuteTaintManager` is part of `NodeLifycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 

<p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>

### Proposed Controllers

The proposed design refactors `NodeLifecycleController` into two independent controllers, which are managed by `kube-controller-manager`.

1. `NodeLifecycleController` monitors node health and adds `NotReady` and `Unreachable` taints to nodes 
2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction

The existing  kube-controller-manager flag `--controllers` will be used to optionally disable the `NoExecuteTaintManager`.

<p align="center"><image src="new-controllers.png" width=60%></p>

> ### Custom taint-manager

Reviewer: I don't think we should mention this part in the design details, as it's out-of-tree option.

Author: The section and diagram were removed from the KEP.

Opting out from the default `TaintManger`, Kubernetes operators will be able to implement custom logic to handle `NoExecute` taint such as delegating the authority for the termination for some classes of workloads providing high-availability protection guarantees. They can also scale out the usage of `NoExcute` to more use-cases than `NotReady` nodes (e.g. unhealthy network, etc.).

<p align="center"><image src="custom-taint-manager.png" width=90%></p>

### Implementation

A new `NodeLifeCycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.

> A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 

Reviewer: We shouldn't say "a new flag", do we? as it will be a top-level controller controlled by `--controllers`.

```suggestion
A new `NoExecuteTaintManager` is created as a first-class controller managed by `kube-controller-manager`. Its implementation is based on the current taint-manager `controller/nodelifecycle/taint-manager.go`.
```

`// NoExecuteTaintManagerConfiguration contains elements describing NoExecuteTaintManagerConfiguration`
`type NoExecuteTaintManagerConfiguration struct {`
`// enables the default NoExecuteTaintManager. MUST be synced with the`
`// corresponding flag of the kube-apiserver.`
`EnableNoExecuteTaintManager bool`
`  ...`

> `}`

Reviewer: Could you reformat this section?

The creation and startup of the default taint-manager is performed by `kube-controller-manager` . 

`func startNoExecuteTaintManager(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {`
`    if !controllerContext.ComponentConfig.NoExecuteTaintManager.EnableNoExecuteTaintManager {
        return nil, false, nil

>     }`

Reviewer: Please format this snippet.

> ```

Reviewer: ```suggestion
```go
```

taintManager, err = taintmanager.NewNoExecuteTaintManager(ctx, kubeClient, nc.podLister, nc.nodeLister, nc.getPodsAssignedToNode)
        nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
            AddFunc: controllerutil.CreateAddNodeHandler(func(node *v1.Node) error {
                nc.taintManager.NodeUpdated(nil, node)
                return nil
            }),
            UpdateFunc: controllerutil.CreateUpdateNodeHandler(func(oldNode, newNode *v1.Node) error {
                nc.taintManager.NodeUpdated(oldNode, newNode)
                return nil
            }),
            DeleteFunc: controllerutil.CreateDeleteNodeHandler(func(node *v1.Node) error {
                nc.taintManager.NodeUpdated(node, nil)
                return nil
            }),
        })
    }
        if err != nil {
        return nil, true, fmt.Errorf("failed to start the taint manager: %v", err)
    }
  }  
```

### Test Plan

* Unit and integration tests
    * Add nodes taints
    * Act on `NoExecute` to delete running pods
    * Respect taint tolerations 
* Verify passing the existing E2E and conformance tests.
    * Taint a node 

>     * Evict the correct set of running pods

Reviewer: You may need to detail the tests section. There is a updated version requiring documenting the current code coverage, etc. You can refer to other latest KEP for references.

Author: How about 

* Unit tests
    * Enable `TaintManager`
    * Disable `TaintManager`
    * Add `NoExecute` taints
    * Act on `NoExecute` to delete running pods
    * Respect taint tolerations
* Integration tests
    * Enable `TaintManager`, add `NoExecute` taints to evict the correct set of running pods on nodes
    * Disable `TaintManager`, add `NoExecute` has no impact on running pods
* E2E tests
    * Verify passing the existing E2E and conformance tests.

### Graduation Criteria

#### Alpha

> * Support `kube-controller-manaager` flag `controllers=-taint-manager` to disable the default taint-manager.

Reviewer: Is the `-` semantics really what we're going to implement? I'd assume disabling a default controller can be achieved by providing a full list of controller names.

If this is a goal of this KEP, I highly recommend to document it in the goals section.

Author: Yes, it's already listed in the goal section.

Reviewer 8: I guess this is side effect of making taint-manager a separate controller. This is not the goal?

Author: By decoupling the taint-manager from nodelifecyclecontroller, we can just use existing controller-manager flag to disable the default taint-manager instead of using a separate flag. I agree it's not the main goal. I'll remove it.
